subprojects/docs/src/docs/userguide/five_things.adoc

```diff
@@ -0,0 +1,126 @@
+
+As a general rule of thumb, everything in a build script happens during the configuration phase unless it defines a task action. There are some other exceptions, such as link:{javadocPath}/org/gradle/api/execution/TaskExecutionGraph.html#whenReady-org.gradle.api.Action-[TaskExecutionGraph.whenReady(Action)], but the documentation is clear on their behavior.
+
+You can learn more in the <<build_lifecycle#build_lifecycle,Build Lifecycle>> chapter.
+
+== 4. Gradle is extensible in more ways than one
+
+It would be great if you could build your project using only the build logic bundled with Gradle, but that's rarely possible. Most builds have some special requirements that mean you need to add custom build logic.
+
+Gradle provides several mechanisms that allow you to extend it, from the very simple to the very powerful:
+
+ 1. Extra properties on <<writing_build_scripts#sec:extra_properties,projects>> and <<writing_build_scripts#sec:extra_task_properties,tasks>>.
++
+These allows you to add your own properties to a project or task that you can then use from your own custom actions or any other build logic. Extra properties can even be applied to tasks that aren't explicitly created by you, such as those created by Gradle's core plugins.
+
+ 2. Custom task actions.
++
+You can attach custom build logic that executes before or after a task via the link:{groovyDslPath}/org.gradle.api.Task.html#org.gradle.api.Task:doFirst(org.gradle.api.Action)[Task.doFirst()] and link:{groovyDslPath}/org.gradle.api.Task.html#org.gradle.api.Task:doLast(org.gradle.api.Action)[Task.doLast()] methods.
+
+ 3. <<custom_tasks,Custom task types>>.
++
+When you want the build to do some work that an existing task can't do, you can simply write your own task type. It's typically best to put the source file for a custom task type in the <<organizing_gradle_projects#sec:build_sources,_buildSrc_>> directory or in a packaged plugin. Then you can use the custom task type just like any of the Gradle-provided ones.
+
+ 4. Custom conventions.
++
+Conventions are a powerful way to simplify builds so that users can understand and use them more easily. This can be seen with builds that use standard project structures and naming conventions, such as <<building_java_projects#building_java_projects,Java builds>>. You can write your own plugins that provide conventions — they just need to configure default values for the relevant aspects of a build.
+
+ 5. {guidesUrl}/implementing-gradle-plugins/#modeling_dsl_like_apis[A custom model].
++
+Gradle allows you to introduce new concepts into a build beyond tasks, files and dependency configurations. You can see this with most language plugins, which add the concept of <<building_java_projects#sec:java_source_sets,_source sets_>> to a build. Appropriate modeling of a build process can greatly improve a build's ease of use and its efficiency. But new model elements should be used sparingly to avoid complex and incomprehensible builds.
```

Comment:
I don't agree with the last sentence. Most incomprehensible builds are due to a lock of modeling, not too much.

Reply:
So what I'm trying to get at is the classic problem of introducing too many abstractions via extensions or other means. I'm reminded of LinkedIn's Ant build which was cleanly done, but you basically had to learn a new build tool due to the abstractions it introduced.

Perhaps it's not worth adding the qualifier here, but rather focus on new abstractions that provide real value.

Reply:
The Android plugin is a great example of an abstraction that's basically a "new build tool" on top of Gradle. It introduces build types, flavors, variants, transformations etc. But at the same time it hides a lot of other things from users and makes it easier to integrate logic at a higher level. I think this is a good approach if your domain is complex enough to warrant the cost.

Reply:
I've removed the last sentence as it's veering into best practice, which isn't really appropriate for this page and would require substantial coverage. I think the "Appropriate modeling" is a sufficient qualifier here.